Search SDK: Adding arm-search spec for 2015-08-19 API version

bitsforthought opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 
### PR information
- [X] The title of the PR is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [X] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [X] If applicable, the PR references the bug/issue that it fixes.
- [X] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).
### Quality of Swagger
- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [X] My spec meets the review criteria:
  - [X] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [X] Validation errors from the [Linter extension for VS Code](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/linter.md) have all been fixed for this spec. (**Note:** for large, previously checked in specs, there will likely be many errors shown. Please contact our team so we can set a timeframe for fixing these errors if your PR is not going to address them).
  - [X] The spec follows the patterns described in the [Swagger good patterns](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md) document unless the service API makes this impossible.

FYI @chaosrealm @seansaleh @HeidiSteen @EvanBoyle @jack4it

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/azure/azure-rest-api-specs/637)

<!-- Reviewable:end -->

bitsforthought committed:
Search SDK: New Swagger spec for Search Management API 2015-08-19

This spec covers all functionality of the Azure Search Management REST API
version 2015-08-19, including key management and search service management,
provisioning, and scaling.

bitsforthought committed:
Search SDK: Updating Management SDK surface area, part 1

This is part 1 of a final pass over the Swagger spec for arm-search. Types of
issues that are addressed include:

  - Consistency of parameter names (serviceName -> searchServiceName).
  - Consistency of method names with established Azure conventions
    (List -> ListByResourceGroup, etc.).
  - Added the ability to pass in x-ms-client-request-id (test coverage still
    pending).
  - Correctly modeled error cases for all key management operations.
  - Enabled paging for List operations (although the Search RP doesn't support
    paging, we at least benefit from the improved client programming model).

Part 2 will address the same sorts of issues in Search service management
operations.

bitsforthought committed:
Search SDK: Part 2 of fleshing out Managment plane documentation

bitsforthought committed:
Search SDK: Updating max partition count for S3 HD

hovsepm commented on arm-search/2015-08-19/swagger/search.json:
```diff
@@ -0,0 +1,894 @@
+        ],
+        "x-ms-pageable": {
+          "nextLinkName": null
+        },
```
Listing services inside resource group will never return more than one page of results?

brjohnstmsft commented:
FYI @tbombach

brjohnstmsft commented:
We'll be replacing the externalDocs links to MSDN as soon as we have new content to replace it.

bitsforthought committed:
Search: Replacing MSDN links with aka.ms links

The aka.ms links still point to MSDN, but they'll be updated to
docs.microsoft.com after the migration.

tbombach commented on arm-search/2015-08-19/swagger/search.json:
```diff
@@ -0,0 +1,888 @@
+      ],
+      "properties": {
+        "name": {
+          "description": "The Search service name to validate. Search service names must only contain lowercase letters, digits or dashes, cannot use dash as the first two or last one characters, cannot contain consecutive dashes, and must be between 2 and 60 characters in length.",
```
It's nice that these constraints are described, but should they also be JSON constraints? The regex pattern might be complicated to do, but the min/max string constraints wouldn't be difficult.

tbombach replied:
The benefit is that validation would fail before a call is made the service.

The constraints (including regex pattern) get turned into checks in the `Validate()` method: https://github.com/Azure/azure-sdk-for-net/blob/04d78e2101f7d040aaa97481757a2fb69fffe9c9/src/Search/Microsoft.Azure.Search/GeneratedSearchService/Models/StandardAnalyzer.cs#L65

That get called when making requests: https://github.com/Azure/azure-sdk-for-net/blob/04d78e2101f7d040aaa97481757a2fb69fffe9c9/src/Search/Microsoft.Azure.Search/GeneratedSearchService/DataSourcesOperations.cs#L102

That way a call would fail fast with a [ValidationException](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/src/ClientRuntime/Microsoft.Rest.ClientRuntime/ValidationException.cs), rather than making an actual call to the service and waiting for a response. The goal is to do client-side validation to the best of our ability, even though there's already server-side validation. It's also easier for a user writing code that uses the client - the response from the service is probably some variation of a 4xx Bad Request, while a ValidationException contains the regex pattern that failed and the target property that failed the pattern. If you rely on the response, then the customer code has to parse the message on the CloudError to find the problem.

brjohnstmsft replied:
This seems like solving the problem at the wrong layer. It introduces unnecessary coupling between the service and the client. Consider versioning -- What if we discover a bug in the regex and fix it in our RP, but forget to fix it in the client? Or maybe we fix it in both places, but some customers are still using an older client with the wrong regex. This kind of coupling seems like too high a price to pay just to shave off some latency in an error case.

Regarding usability, here is the experience of providing an invalid service name today:

```csharp
[Fact]
public void CreateServiceWithInvalidNameGivesUsefulMessage()
{
    Run(() =>
    {
        SearchManagementClient searchMgmt = GetSearchManagementClient();
        SearchService service = DefineServiceWithSku(SkuName.Free);

        CloudException e =
            Assert.Throws<CloudException>(() => searchMgmt.Services.CreateOrUpdate(Data.ResourceGroupName, InvalidServiceName, service));
        Assert.Equal("Service name '----badname' is invalid: Service name must only contain lowercase letters, digits or dashes, cannot start or end with or contain consecutive dashes and is limited to 60 characters.", e.Message);
    });
}
```

If need be, we can include the regex in the error message, but for most users this won't make much sense.

tbombach commented on arm-search/2015-08-19/swagger/search.json:
```diff
@@ -0,0 +1,888 @@
+          ],
+          "x-ms-enum": {
+            "name": "SkuName",
+            "modelAsString": false
```
One note with `modelAsString: false` is if a new value is ever added, the client will be unable to handle it. When trying to deserialize the `name` property, it will be unable to map any values not listed here to a correct enum value (for languages that use real enums).

brjohnstmsft replied:
Good point. One thing I don't like about modelAsString though is that it breaks Intellisense. In our data plane SDK, I created an "extensible enum" pattern to handle this, but I haven't had time to try to fold it back into AutoRest.

How about this... Since we don't plan to add any SKUs in the near future, let's leave it like this for now, and I'll use the addition of a new SKU as a forcing function add extensible enums to AutoRest. Sound good?

tbombach replied:
Yes, I think that's our currently only supported way of having the desired Intellisense experience for a set of known values. Let's leave it that way for now with the knowledge that a major version bump for SDKs would be required once SKU is expanded (or the extensible enums that you mentioned).

tbombach reviewed (changes requested):
The JSON constraints aren't a "must-have" for merging (especially not a regex pattern that covers the rules for dashes).

The only question I have before merging is around use of enums - is it fine that the client will be unable to handle new values for that property? It would take a major version bump in the SDKs to support a new value (since enum additions are considered breaking changes in strongly typed languages). If you want to minimize breaking changes, then modeling as string will do that.

brjohnstmsft commented:
I decided to add some JSON examples to the spec... Please hold off merging for now. Thanks!

bitsforthought committed:
Search: Adding examples to arm-search spec

brjohnstmsft commented:
@HeidiSteen I ended up adding the examples after all, since it's possible to have different examples for the request & response.

@tbombach I'm ready for this to be merged now, assuming you agree with my comments above.

tbombach commented:
Thanks for the update @brjohnstmsft. The enum plan makes sense. I left another comment re: the regex constraint.

brjohnstmsft commented:
@tbombach Replied to your comment above.

AutorestCI commented:
No modification for Ruby

AutorestCI commented:
No modification for NodeJS

AutorestCI commented:
No modification for Python

seansaleh commented on arm-search/2015-08-19/swagger/search.json:
```diff
@@ -548,7 +682,11 @@
-      "description": "Input of check name availability API."
+      "description": "Input of check name availability API.",
+      "example": {
+        "name": "azs-test",  
```
Review opened to change it
